variable/test/trigonometry_test.cpp

```diff
@@ -3,157 +3,190 @@
+    auto values_in_rad = var.values<double>();
+    std::transform(values_in_rad.begin(), values_in_rad.end(),
+                   back_inserter(res),
+                   [&op](double x_in_rad) { return op(x_in_rad); });
+    return makeVariable<double>(Dims{Dim::X}, Shape{res.size()}, Values(res));
+  }
+};
+
+TEST_F(VariableTrigonometryTest, sin_rad) {
+  const auto var = input_in_rad();
+  EXPECT_EQ(sin(var), expected_for_op(std::sin));
+  EXPECT_EQ(var, input_in_rad());
 }
 
-TEST(VariableTrigonometryTest, sin_deg) {
-  const auto var = makeVariable<double>(Values{180.0}, Unit{units::deg});
-  EXPECT_EQ(sin(var), makeVariable<double>(Values{std::sin(pi<double>)}));
+TEST_F(VariableTrigonometryTest, sin_deg) {
+  const auto var = input_in_deg();
+  EXPECT_EQ(sin(var), expected_for_op(std::sin));
+  EXPECT_EQ(var, input_in_deg());
 }
 
-TEST(Variable, sin_move_rad) {
-  auto var = makeVariable<double>(Values{pi<double>}, Unit{units::rad});
+TEST_F(VariableTrigonometryTest, sin_move_rad) {
+  auto var = input_in_rad();
   const auto ptr = var.values<double>().data();
   auto out = sin(std::move(var));
-  EXPECT_EQ(out, makeVariable<double>(Values{std::sin(pi<double>)}));
+  EXPECT_EQ(out, expected_for_op(std::sin));
   EXPECT_EQ(out.values<double>().data(), ptr);
 }
 
-TEST(Variable, sin_move_deg) {
-  auto var = makeVariable<double>(Values{180.0}, Unit{units::deg});
+TEST_F(VariableTrigonometryTest, sin_move_deg) {
+  auto var = input_in_deg();
   const auto ptr = var.values<double>().data();
   auto out = sin(std::move(var));
-  EXPECT_EQ(out, makeVariable<double>(Values{std::sin(pi<double>)}));
+  EXPECT_EQ(out, expected_for_op(std::sin));
   EXPECT_EQ(out.values<double>().data(), ptr);
 }
 
-TEST(Variable, sin_out_arg_rad) {
-  auto x = makeVariable<double>(Dims{Dim::X}, Shape{2}, Values{pi<double>, 0.0},
-                                Unit{units::rad});
-  auto out = makeVariable<double>(Values{0.0});
-  auto &view = sin(x.slice({Dim::X, 0}), out);
+TEST_F(VariableTrigonometryTest, sin_out_arg_rad) {
+  const auto in = input_in_rad();
+  auto out = special_like(in, FillValue::ZeroNotBool);
+  auto &view = sin(in, out);
 
-  EXPECT_EQ(out, makeVariable<double>(Values{std::sin(pi<double>)}));
+  EXPECT_EQ(out, expected_for_op(std::sin));
   EXPECT_EQ(&view, &out);
+  EXPECT_EQ(in, input_in_rad());
 }
 
-TEST(Variable, sin_out_arg_deg) {
-  auto x = makeVariable<double>(Dims{Dim::X}, Shape{2}, Values{180.0, 0.0},
-                                Unit{units::deg});
-  auto out = makeVariable<double>(Values{0.0});
-  auto &view = sin(x.slice({Dim::X, 0}), out);
+TEST_F(VariableTrigonometryTest, sin_out_arg_deg) {
+  const auto in = input_in_deg();
+  auto out = special_like(in, FillValue::ZeroNotBool);
+  auto &view = sin(in, out);
 
-  EXPECT_EQ(out, makeVariable<double>(Values{std::sin(pi<double>)}));
+  EXPECT_EQ(out, expected_for_op(std::sin));
   EXPECT_EQ(&view, &out);
+  EXPECT_EQ(in, input_in_deg());
```

Comment:
Now if someone ever has the smart idea of turning `input_in_deg()` into a Meyer's singleton, always returning a variable referencing the same buffer, this will not test anything. Maybe better to check `EXPECT_NE(in, out)`?

Reply:
Not sure that test is better because it still allows for modification of `in`.